Refactor Form Component

bafonins opened the pull request:
<!--
Thanks for submitting a pull request. Please fill the template below,
otherwise we will not be able to process this pull request.
-->

#### Summary
<!--
A short summary, referencing related issues:
Closes #0000, References #0000, etc.
-->

Closes https://github.com/TheThingsNetwork/lorawan-stack/issues/582
This PR introduces a new implementation of the `<Form />` and `<Field />` components, as well other util components. Besides that, I refactored some of the forms to show how the new implementation can be used.

#### Changes
<!-- What are the changes made in this pull request? -->

- Add the new `<Form />` component
- Add the new `<Field />` component
- Refactor `<Checkbox />` and introduce the `<CheckboxGroup />` component
- Refactor `<Radio />` and introduce the `<RadioGroup />` component
- Add adjustments to `<Input />` and `<Select />`
- Fix some issues related to form handling on different pages
- Refactor `<Field />` storybook

#### Notes for Reviewers
<!--
NOTE: This section is optional.

Motivate briefly why it is implemented this way, if that deviates from the
implementation proposal in the referenced issues.
- How should your reviewers approach this pull request?
- @mention reviewers with special requests or questions for them
-->

With this approach we have much more control over adding new functionality to the form.
Now, any field can be added a form component via the `<Field />` component as long as it complies to the props passed by the form context.
I removed `mapErrorsToFields` functionality from the form, because there is no clear way to implement this properly.

kschiffer commented on pkg/webui/components/new-form/field/index.js:
```diff
@@ -0,0 +1,210 @@
+    warning: '',
+    description: '',
+    required: false,
+  }
```
Should we default `component` to `Input`, to stay in line with the old form's api?

kschiffer commented on pkg/webui/components/new-form/field/index.js:
```diff
@@ -0,0 +1,210 @@
+    const showDescription = !showError && !showWarning && hasDescription
+
+    const fieldMessage = showError ? (
+      <div className={style.messages}>
```
We could try to render this div by default with a fixed `min-height`, so that the validation messages will not result in a jump.

kschiffer commented on pkg/webui/console/components/gateway-data-form/index.js:
```diff
@@ -61,59 +63,58 @@ class GatewayDataForm extends React.Component {
         initialValues={initialValues}
         validationSchema={validationSchema}
-        mapErrorsToFields={mapErrorsToFields}
         horizontal
```
```suggestion
```

kschiffer commented on pkg/webui/console/containers/device-data-form/index.js:
```diff
@@ -218,47 +232,45 @@ class DeviceDataForm extends Component {
-          id_taken: 'application_id',
-          identifiers: 'application_id',
-        }}
         horizontal
```
```suggestion
```

kschiffer commented on pkg/webui/console/containers/device-data-form/index.js:
```diff
@@ -290,38 +302,36 @@ class DeviceDataForm extends Component {
           horizontal
```
```suggestion
```

Should not be necessary anymore to extra pass this to the Select field

bafonins commented:
> Radio buttons have no labels in the story

fixed

> Text fields that have messages (react-intl) as title will show [object Object] as tooltip

fixed

> The submit button becomes valid as soon as any input occurred, even though it might not be valid

this is expected behaviour. initially the button is disabled because the form is pristine

> What would be the plan for replacing (and removing) the old form component?

after we agree on the implementation I will refactor the rest of the forms and remove the old form/field within this pr

> Can we use the opportunity to avoid layout jumps on error messages?

dont want to spend time on that now

> Field Groups story renders checkboxes incorectly:

didnt refactor this part. tried to keep changes as small as possible.

kschiffer commented on pkg/webui/components/radio-button/radio-button.styl:
```diff
@@ -12,58 +12,81 @@
-.container
-  display: flex
+.wrapper
+  cursor: pointer
```
When in a horizontal group, the wrapper needs more spacing to the right.
![image](https://user-images.githubusercontent.com/5710611/59428189-e4a74600-8ddc-11e9-9fe7-ead9ca948238.png)

kschiffer commented on pkg/webui/components/submit-button/index.js:
```diff
@@ -0,0 +1,60 @@
+
+SubmitButton.defaultProps = {
+  disabled: false,
+  icon: undefined,
```
```suggestion
```

kschiffer commented on pkg/webui/oauth/views/login/index.js:
```diff
@@ -92,20 +102,26 @@ export default class OAuth extends React.PureComponent {
               initialValues={initialValues}
               error={this.state.error}
-              submitEnabledWhenInvalid
+              validationSchema={validationSchema}
```
```suggestion
              validationSchema={validationSchema}
              horizontal={false}
```

kschiffer commented on pkg/webui/oauth/views/create-account/index.js:
```diff
@@ -133,61 +142,64 @@ export default class CreateAccount extends React.PureComponent {
+              initialValues={initialValues}
+              error={error}
+              info={info}
+              validationSchema={validationSchema}
```
```suggestion
              validationSchema={validationSchema}
              horizontal={false}
```

kschiffer commented on pkg/webui/oauth/views/create-account/index.js:
```diff
@@ -133,61 +142,64 @@ export default class CreateAccount extends React.PureComponent {
-      </div>
+      <Container className={style.fullHeight}>
+        <Row justify="center" align="center" className={style.fullHeight}>
+          <Col lg={8} md={12}>
```
```suggestion
          <Col lg={5} md={8} sm={12}>
```

kschiffer commented on pkg/webui/oauth/views/create-account/create-account.styl:
```diff
@@ -12,36 +12,11 @@
-    display: block
+.title
+  h1()
+  text-align: center
```
```suggestion
```

kschiffer commented on pkg/webui/components/input/index.js:
```diff
@@ -91,6 +116,16 @@ class Input extends React.Component {
+    }
+
+    let inputTitle = title
+    if (typeof placeholder === 'object') {
```
```suggestion
    if (typeof title === 'object') {
```

kschiffer commented on pkg/webui/components/checkbox/group/group.styl:
```diff
@@ -12,44 +12,15 @@
-  display: flex
-  flex-wrap: wrap
+  .group-checkbox:not(:last-child)
+    margin-right: $cs.s
```
```suggestion
    margin-right: $cs.m
```
